http-client/Network/HTTP/Client/Request.hs

```diff
@@ -370,26 +371,32 @@ requestBuilder req Connection {..} = do
     toTriple (RequestBodyStream len stream) = do
         -- See https://github.com/snoyberg/http-client/issues/74 for usage
         -- of flush here.
-        let body = writeStream False stream
+        let body = writeStream (Just . fromIntegral $ len) stream
             -- Don't check for a bad send on the headers themselves.
             -- Ideally, we'd do the same thing for the other request body
             -- types, but it would also introduce a performance hit since
             -- we couldn't merge request headers and bodies together.
             now  = flushHeaders (Just len) >> checkBadSend body
         return (Just len, now, body)
     toTriple (RequestBodyStreamChunked stream) = do
-        let body = writeStream True stream
+        let body = writeStream Nothing stream
             now  = flushHeaders Nothing >> checkBadSend body
         return (Nothing, now, body)
     toTriple (RequestBodyIO mbody) = mbody >>= toTriple
 
-    writeStream isChunked withStream =
-        withStream loop
+    writeStream mlen withStream =
+        withStream (loop 0) 
       where
-        loop stream = do
+        -- When stream is chunked, no length is provided 
+        isChunked = isNothing mlen
+        loop !n stream = do
             bs <- stream
             if S.null bs
-                then when isChunked $ connectionWrite "0\r\n\r\n"
+                then 
+                    if isChunked then connectionWrite "0\r\n\r\n"
+                    -- If not chunked, then length argument is present
+                    -- and should be validated
+                    else unless (fromJust mlen == n) $ throwIO $ WrongRequestBodyStreamSize (fromIntegral . fromJust $ mlen) (fromIntegral n)
```

Comment:
I don't feel comfortable including a partial function like `fromJust` here. Why not just pattern match on `mlen` instead of using `isChunked`?